server/src/main/java/org/elasticsearch/common/bytes/ReleasableBytesReference.java

```diff
@@ -27,17 +27,15 @@
 
     public static final Releasable NO_OP = () -> {};
     private final BytesReference delegate;
-    private final AbstractRefCounted refCounted;
+    protected final AbstractRefCounted refCounted;
 
     public ReleasableBytesReference(BytesReference delegate, Releasable releasable) {
-        this.delegate = delegate;
-        this.refCounted = new RefCountedReleasable(releasable);
+        this(delegate, new RefCountedReleasable(releasable));
     }
 
-    private ReleasableBytesReference(BytesReference delegate, AbstractRefCounted refCounted) {
+    public ReleasableBytesReference(BytesReference delegate, AbstractRefCounted refCounted) {
         this.delegate = delegate;
         this.refCounted = refCounted;
-        refCounted.incRef();
```

Comment:
The reason I did it this way around was that `delegate.slice()` could throw in theory (out of bounds or other future stuff depending on what `delegate` will be like). So if I increment ref count first, I'll have to add the whole `try-finally` dance to release the reference in case that happens => this seemed less noisy + it didn't seem so bad to get here with ref count `0` as this situation would be caught by other assertions anyway?

Reply:
This seems safe and clean?

```
        final BytesReference slice = delegate.slice(from, length);
        refCounted.incRef();
        return new ReleasableBytesReference(slice, refCounted);
```

Reply:
:facepalm: sorry for being slow here => that's perfect will do :)